File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
void generateParenthesis(int n , vector<string>&ans , string &temp , int left , int right){
   
if(left==n && right==n){
    ans.push_back(temp);
   
    return;
}
 
 
  if(left<n){
      temp+='(';
       generateParenthesis(n , ans , temp,left+1,right);
       temp.pop_back();
       
  } 
 
 if(right<left && right<n){
          
          temp+=')';
          generateParenthesis(n , ans , temp ,left,right+1);
          temp.pop_back();
 } 


}


    vector<string> generateParenthesis(int n) {
        vector<string>ans;
        string temp="";
        

            generateParenthesis(n , ans , temp ,0,0);

            return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */

function generate(ans,n,string,left,right){
if(n==left && n==right){
    ans.push(string);
    return;
}

if(left<n) {
    
    generate(ans , n,string+"(",left+1,right);
}
if(right<left){
     
    generate(ans , n,string+")",left,right+1);
}



}
var generateParenthesis = function(n) {
    let string = "";
    let ans = [];


    generate(ans,n,string,0,0);


    return ans;
};